File: 988.py
class Solution:
    def smallestFromLeaf(self, root: Optional[TreeNode]) -> str:
        minStr = '{'
        stack = [(root, '')]
        while len(stack) != 0:
            node, suff = stack.pop()
            suff += chr(97 + node.val)
            if node.left is None and node.right is None:
                minStr = min(minStr, suff[::-1])
            else:
                if node.left is not None:
                    stack.append((node.left, suff))
                if node.right is not None:
                    stack.append((node.right, suff))
        return minStr

File: 988.cpp
struct Solution {
    void minStr(TreeNode *node, string &currSuff, string &currMin){
        currSuff += static_cast<char>(node->val + 'a');
        if (node->left == nullptr && node->right == nullptr){
            string copy(currSuff);
            reverse(copy.begin(), copy.end());
            currMin = min(currMin, copy);
        } else {
            if (node->left != nullptr){
                minStr(node->left, currSuff, currMin);
            }
            if (node->right != nullptr){
                minStr(node->right, currSuff, currMin);
            }
        }
        currSuff.pop_back();
    }
    string smallestFromLeaf(TreeNode* root) {
        string currSuff = "", currMin = "{";
        minStr(root, currSuff, currMin);
        return currMin;
    }
};
